Sources/TensorFlow/Core/Utilities.swift

```diff
@@ -167,3 +167,153 @@ extension UnsafeMutablePointer where Pointee == CTensorHandle? {
+        accumulated += lengthBytes
+
+        // Step 3: Process the array bytes.
+        var accumulatedHash = SIMD8<UInt64>(
+            0x6a09e667f3bcc908, 0xbb67ae8584caa73b, 0x3c6ef372fe94f82b, 0xa54ff53a5f1d36f1, 
+            0x510e527fade682d1, 0x9b05688c2b3e6c1f, 0x1f83d9abfb41bd6b, 0x5be0cd19137e2179)
+        var index = 0
+        while index < accumulated.count {
+            let chunk = accumulated[index..<(index + blockSize)]
+            index += blockSize
+
+            // Break chunk into sixteen 64-bit words w[j], 0 ≤ j ≤ 15, in big-endian format.
+            // Extend the sixteen 64-bit words into eighty 64-bit words:
+            var w = [UInt64](repeating: 0, count: k.count)
+            for x in k.indices {
+                switch x {
+                case 0...15:
+                    let start = chunk.startIndex.advanced(by: x * 8)
+                    w[x] = UInt64(bytes: chunk, startingAt: start)
+                    break
+                default:
+                    let s0Term0 = ((w[x - 15] >> 1 ^ w[x - 15]) >> 6 ^ w[x - 15]) >> 1
+                    let s0Term1 = (w[x - 15] << 7 ^ w[x - 15]) << 56
+                    let s0 = s0Term0 ^ s0Term1
+                    let s1Term0 = ((w[x - 2] >> 42 ^ w[x - 2]) >> 13 ^ w[x - 2]) >> 6
+                    let s1Term1 = (w[x - 2] << 42 ^ w[x - 2]) << 3
+                    let s1 = s1Term0 ^ s1Term1
+                    w[x] = w[x - 16] &+ s0 &+ w[x - 7] &+ s1
+                    break
+                }
+            }
+
+            var hashCopy = accumulatedHash
+            for j in k.indices {
+                let s0Term0 = ((hashCopy[0] >> 5 ^ hashCopy[0]) >> 6 ^ hashCopy[0]) >> 28
+                let s0Term1 = ((hashCopy[0] << 6 ^ hashCopy[0]) << 5 ^ hashCopy[0]) << 25
+                let s0 = s0Term0 ^ s0Term1
+                let s1Term0 = ((hashCopy[4] >> 23 ^ hashCopy[4]) >> 4 ^ hashCopy[4]) >> 14
+                let s1Term1 = ((hashCopy[4] << 4 ^ hashCopy[4]) << 23 ^ hashCopy[4]) << 23
+                let s1 = s1Term0 ^ s1Term1
+                let maj = (hashCopy[0] & hashCopy[1]) ^ 
+                    (hashCopy[0] & hashCopy[2]) ^
+                    (hashCopy[1] & hashCopy[2])
+                let t2 = s0 &+ maj
+                let ch = (hashCopy[4] & hashCopy[5]) ^ (~hashCopy[4] & hashCopy[6])
+                let t1 = hashCopy[7] &+ s1 &+ ch &+ k[j] &+ w[j]
+                hashCopy[7] = hashCopy[6]
+                hashCopy[6] = hashCopy[5]
+                hashCopy[5] = hashCopy[4]
+                hashCopy[4] = hashCopy[3] &+ t1
+                hashCopy[3] = hashCopy[2]
+                hashCopy[2] = hashCopy[1]
+                hashCopy[1] = hashCopy[0]
+                hashCopy[0] = t1 &+ t2
+            }
+            accumulatedHash &+= hashCopy
+        }
+
+        // Step 4: Return the computed hash.
+        var result = SIMD64<UInt8>()
+        // var result = [UInt8](repeating: 0, count: digestLength)
```

Comment:
Could this be removed?

Reply:
Yes sorry it's still a WIP because I'm looking into whether I can vectorize some more parts. If I can't I'll just go ahead and merge this a bit later.

By the way, should we make the `sha512()` extension public? It's a nice implementation of SHA512 that's missing from the standard library anyway and can come in handy. I think it's fine either way so it's your call.

Reply:
@rxwei what do you think about making `sha512()` public? This is ready to merge otherwise as I don't see a simple way to vectorize the rest.

I also have an `sha1()` implementation for Swift Retro that I've been building and I could open a separate PR adding a `Hashing.swift` utilities file with both implementations.

Reply:
Let's make it internal for now. Would be great to bring this up on the Swift forum.